### PR TITLE
[SPARK-31975][SQL] Show AnalysisException when WindowFunction is used without WindowExpression

### PR DESCRIPTION
**Pull request by ulysses-you**
### What changes were proposed in this pull request?

Add WindowFunction check at `CheckAnalysis`.

### Why are the changes needed?
Provide friendly error msg.

**BEFORE**
```scala
scala> sql("select rank() from values(1)").show
java.lang.UnsupportedOperationException: Cannot generate code for expression: rank()
```

**AFTER**
```scala
scala> sql("select rank() from values(1)").show
org.apache.spark.sql.AnalysisException: Window function rank() requires an OVER clause.;;
Project [rank() AS RANK()#3]
+- LocalRelation [col1#2]
```

### Does this PR introduce _any_ user-facing change?

Yes, user wiill be given a better error msg.

### How was this patch tested?

Pass the newly added UT.

### CODE REVIEW

**Comment by wangyum** [comment]
retest this please.

File: sql/catalyst/src/main/scala/org/apache/spark/sql/catalyst/analysis/CheckAnalysis.scala [hunk 1]
```diff
@@ -158,6 +158,9 @@ trait CheckAnalysis extends PredicateHelper {
           case g: GroupingID =>
             failAnalysis("grouping_id() can only be used with GroupingSets/Cube/Rollup")
 
+          case Alias(w: WindowFunction, _) =>
+            failAnalysis(s"Expression '$w' not supported without a window function.")
```
Comment (wangyum) [inline sql/catalyst/src/main/scala/org/apache/spark/sql/catalyst/analysis/CheckAnalysis.scala:162]: How about `window function call requires an OVER clause`?

File: sql/catalyst/src/main/scala/org/apache/spark/sql/catalyst/analysis/CheckAnalysis.scala [hunk 2]
```diff
@@ -158,6 +158,9 @@ trait CheckAnalysis extends PredicateHelper {
           case g: GroupingID =>
             failAnalysis("grouping_id() can only be used with GroupingSets/Cube/Rollup")
 
+          case Alias(w: WindowFunction, _) =>
+            failAnalysis(s"Window function '$w' call requires an OVER clause.")
```
Comment (wangyum) [inline sql/catalyst/src/main/scala/org/apache/spark/sql/catalyst/analysis/CheckAnalysis.scala:162]: Sorry, seems we need to remove single quotation marks:
https://github.com/apache/spark/blob/b2011a295bd78b3693a516e049e90250366b8f52/sql/catalyst/src/main/scala/org/apache/spark/sql/catalyst/analysis/Analyzer.scala#L2734

**Comment by wangyum** [comment]
cc @jiangxb1987

**Comment by ulysses-you** [comment]
also cc @HyukjinKwon  @cloud-fan  a minor change.

File: sql/catalyst/src/main/scala/org/apache/spark/sql/catalyst/analysis/CheckAnalysis.scala [hunk 3]
```diff
@@ -158,6 +158,9 @@ trait CheckAnalysis extends PredicateHelper {
           case g: GroupingID =>
             failAnalysis("grouping_id() can only be used with GroupingSets/Cube/Rollup")
 
+          case Alias(w: WindowFunction, _) =>
+            failAnalysis(s"Window function $w call requires an OVER clause.")
```
Comment (cloud-fan) [inline sql/catalyst/src/main/scala/org/apache/spark/sql/catalyst/analysis/CheckAnalysis.scala:162]: nit: we can remove `call`

**Review by cloud-fan — APPROVED** [review]
LGTM, please resolve the conflicts.

File: sql/catalyst/src/main/scala/org/apache/spark/sql/catalyst/analysis/CheckAnalysis.scala (hunk 3 above)
Comment (ulysses-you) [reply sql/catalyst/src/main/scala/org/apache/spark/sql/catalyst/analysis/CheckAnalysis.scala]: removed.

File: sql/catalyst/src/main/scala/org/apache/spark/sql/catalyst/analysis/CheckAnalysis.scala [hunk 4]
```diff
@@ -158,6 +158,9 @@ trait CheckAnalysis extends PredicateHelper {
           case g: GroupingID =>
             failAnalysis("grouping_id() can only be used with GroupingSets/Cube/Rollup")
 
+          case Alias(w: WindowFunction, _) =>
```
Comment (cloud-fan) [inline sql/catalyst/src/main/scala/org/apache/spark/sql/catalyst/analysis/CheckAnalysis.scala:161]: BTW why do we have to match the `Alias`?

Comment (ulysses-you) [reply sql/catalyst/src/main/scala/org/apache/spark/sql/catalyst/analysis/CheckAnalysis.scala:161]: The right way is `Alias(WindowExpression(w: WindowFunction, _),  _)` and without `over` it is `Alias(w: WindowFunction, _)`.

Comment (cloud-fan) [reply sql/catalyst/src/main/scala/org/apache/spark/sql/catalyst/analysis/CheckAnalysis.scala:161]: then the check doesn't work for cases like `rank() + 1`?

Comment (ulysses-you) [reply sql/catalyst/src/main/scala/org/apache/spark/sql/catalyst/analysis/CheckAnalysis.scala:161]: good catch ! we need to check all expression.

**Review by cloud-fan — APPROVED** [review]

**Comment by cloud-fan** [comment]
retest this please

File: sql/catalyst/src/test/scala/org/apache/spark/sql/catalyst/analysis/AnalysisSuite.scala [hunk 5]
```diff
@@ -884,4 +884,15 @@ class AnalysisSuite extends AnalysisTest with Matchers {
       Seq("Intersect can only be performed on tables with the compatible column types. " +
         "timestamp <> double at the second column of the second table"))
   }
+
+  test("throw user facing error when use WindowFunction directly") {
```
Comment (HyukjinKwon) [inline sql/catalyst/src/test/scala/org/apache/spark/sql/catalyst/analysis/AnalysisSuite.scala:888]: Shall we add a JIRA prefix here? `SPARK-31975: ...`

**Review by HyukjinKwon — APPROVED** [review]

File: sql/catalyst/src/main/scala/org/apache/spark/sql/catalyst/analysis/CheckAnalysis.scala [hunk 6]
```diff
@@ -158,6 +158,11 @@ trait CheckAnalysis extends PredicateHelper {
           case g: GroupingID =>
             failAnalysis("grouping_id() can only be used with GroupingSets/Cube/Rollup")
 
+          case e: Expression if e.children.exists(_.isInstanceOf[WindowFunction]) &&
+            !e.isInstanceOf[WindowExpression] =>
```
Comment (maropu) [inline sql/catalyst/src/main/scala/org/apache/spark/sql/catalyst/analysis/CheckAnalysis.scala:162]: nit: indent

**Review by maropu — APPROVED** [review]

**Review by dongjoon-hyun — APPROVED** [review]
+1, LGTM. Thank you, @ulysses-you .

BTW, @ulysses-you . I updated the PR title and the description a little.

**Comment by HyukjinKwon** [comment]
retest this please

**Comment by ulysses-you** [comment]
Thank you @dongjoon-hyun , thanks for help test @HyukjinKwon .

**Comment by cloud-fan** [comment]
retest this please

**Comment by cloud-fan** [comment]
retest this please

**Comment by HyukjinKwon** [comment]
retest this please

**Comment by wangyum** [comment]
retest this please.

**Comment by cloud-fan** [comment]
thanks, merging to master!

**Comment by ulysses-you** [comment]
thanks for merging! thanks all!

**Pull request closed without merging**